### PR TITLE
[Model] Add Mage-Flow support

### LINKED ISSUE #5337: [New Model]: microsoft/Mage-Flow
### The model to consider.

t2i: https://huggingface.co/microsoft/Mage-Flow

i2i: https://huggingface.co/microsoft/Mage-Flow-Edit

github: https://github.com/microsoft/Mage

### The closest model vllm-omni already supports.

 Qwen-Image / Qwen-Image-Edit

### What's your difficulty of supporting the model you want?

_No response_

### Use case and motivation

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by BruceLoveDecimal**
# [New Model] Add Mage-Flow support

## Purpose

Add support for **Mage-Flow**, Microsoft's 4B NR-MMDiT for text-to-image generation and multi-reference image editing, and make it run multi-GPU across three parallelism dimensions.

Mage-Flow is a dual-stream MMDiT
image

The closest existing model in vLLM-Omni is **Qwen-Image**. Mage-Flow reuses the same dual-stream block structure and the same tensor/sequence parallel approach, while adding padded request batching, a native Mage VAE. The upstream content-policy gate and Gaussian-Shading watermark are not included.
Closes [#5337](https://github.com/vllm-project/vllm-omni/issues/5337)

### Model Core

Added native Mage-Flow support under `vllm_omni/diffusion/models/mage_flow/`:
- `pipeline_mage_flow.py` — T2I and multi-reference Edit pipeline, prompt encoding via Qwen3-VL, padded request batching, packed CFG, native denoising loop
- `mage_flow_transformer.py` — the NR-MMDiT: dual-stream blocks, per-sample image RoPE, `_sp_plan` for sequence parallelism, checkpoint weight loading with fused-QKV remapping
- `mage_flow_layers.py` — joint attention, TP-aware projections, `MageFlowImageRopePrepare`
- `autoencoder_mage.py` — native Mage VAE
- `prompt_utils.py` — prompt templating, variant defaults, reference-image preprocessing

### Parallelism

| Dimension | Status |
| --- | --- |
| Tensor Parallel | Same structure as Qwen-Image: fused `QKVParallelLinear`, `RowParallelLinear` outputs, column→row FFN, `ReplicatedLinear` modulation |
| CFG Parallel | One guidance branch per rank via the shared `CFGParallelMixin`; degree must be 1 or 2 |
| Sequence Parallel (Ulysses) | Same `_sp_plan` approach as Qwen-Image: shard the image stream, keep text replicated and rejoin it inside attention as joint tensors |
| Pipeline Parallel · VAE Patch Parallel · HSDP | Rejected at init with an explicit error rather than degrading silently |

## Test Results

1 | lake | A serene mountain lake at sunrise, mist over the water, pine trees along the shore, photorealistic
lake_omni
2 | portrait | Close-up portrait of an elderly fisherman, weathered face, soft window light, shallow depth of field, photorealistic
portrait_omni
3 | street | A rainy Tokyo street at night, neon signs reflecting on wet asphalt, people with umbrellas, cinematic
street_omni
4 | still | A still life of citrus fruit and a glass carafe on a linen cloth, morning light, painterly
still_omni
5 | wide | A vast desert canyon at golden hour, layered rock strata, a lone hawk
wide_omni

### e2e Metrics

Hardware: 4× NVIDIA H20 (97 GB, NVLink NV18) · torch 2.11.0+cu130 · vLLM 0.24.0
Config: Mage-Flow-Turbo, 1024×1024, 4 steps, seed 42, single request

Sequence parallelism requires a batch of one, which packed CFG breaks, so the SP rows run at guidance 1.0 against their own single-GPU reference. Comparing them against the guidance-4.0 group would conflate the guidance change with the parallelism effect.

| config | SSIM | PSNR | latency | peak VRAM |
| --- | --- | --- | --- | --- |
| 1 GPU (guidance 4.0) | — | — | 11.96 s | 18288 MB |
| TP=2 | 0.9921 | 37.89 dB | 11.59 s | 15662 MB |
| TP=4 | 0.9911 | 37.68 dB | 11.87 s | 14348 MB |
| CFG=2 | 0.9908 | 37.24 dB | 6.26 s | 18014 MB |
| TP=2 × CFG=2 | 0.9893 | 36.22 dB | 3.99 s | 15644 MB |
| SP=2 × CFG=2 | 0.9889 | 35.51 dB | 1.68 s | 18136 MB |
| 1 GPU (guidance 1.0) | — | — | 1.50 s | 18140 MB |
| SP=2 | 0.9969 | 46.71 dB | 1.30 s | 18136 MB |
| SP=4 | 0.9968 | 46.49 dB | 1.23 s | 18136 MB |

## Docs

- `recipes/microsoft/Mage-Flow.md` — 1 / 2 / 4 GPU configurations with the measured numbers above, and which flag to reach for given a latency / memory / fidelity goal
- Model and serving docs updated; they previously stated multi-GPU execution was unsupported

## Testing

- **E2E**: offline + online serving for both T2I and Edit, plus expansion tests wired into the nightly X2I group
- **Perf**: `tests/dfx/perf/tests/test_mage_flow{,_edit}_vllm_omni.json`
- CI entries in `.buildkite/cuda/test-ready.yml`, `test-merge.yml`, `test-nightly.yml`

### CODE REVIEW

**Review by RuixiangMa — COMMENTED** [review]
@BruceLoveDecimal Thanks for your contribution, left a few comments:
1. recipes/microsoft/Mage-Flow.md is added but not indexed in the Available Recipes table. Every other model recipe has a row.
2. Missing docs/user_guide/diffusion_features.md ImageGen table row
3. PR description lists content_policy.py — upstream content-safety check and Gaussian-Shading watermark under Model Core. The file doesn't exist in the changed files. Update the PR body to match what's shipped.

**Comment by BruceLoveDecimal** [comment]
> @BruceLoveDecimal Thanks for your contribution, left a few comments:
> 
> 1. recipes/microsoft/Mage-Flow.md is added but not indexed in the Available Recipes table. Every other model recipe has a row.
> 2. Missing docs/user_guide/diffusion_features.md ImageGen table row
> 3. PR description lists content_policy.py — upstream content-safety check and Gaussian-Shading watermark under Model Core. The file doesn't exist in the changed files. Update the PR body to match what's shipped.

 Thanks for the review! I've addressed all the comments by updating the PR description, recipes and docs.

File: vllm_omni/diffusion/models/mage_flow/pipeline_mage_flow.py [hunk 1]
```diff
@@ -0,0 +1,983 @@
...
+            raise ValueError("num_inference_steps must be positive")
+        guidance_scale = sampling.guidance_scale if sampling.guidance_scale_provided else defaults.guidance_scale
+        if guidance_scale < 1.0:
+            if not is_dummy_warmup:
+                raise ValueError("Mage-Flow guidance_scale must be at least 1.0")
+            # The engine's warmup request pins guidance_scale=0.0 to keep CFG
+            # off. Mage-Flow spells that same intent as 1.0, so accept it here
+            # rather than failing startup, while real requests stay validated.
+            guidance_scale = 1.0
+        output_type = sampling.output_type or "pil"
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/mage_flow/pipeline_mage_flow.py]: Use self.od_config.output_type when the request omits output_type.

File: vllm_omni/diffusion/models/mage_flow/mage_flow_transformer.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+        else:
+            encoder_attention_mask = torch.ones(
+                encoder_hidden_states.shape[:2],
+                dtype=torch.bool,
+                device=encoder_hidden_states.device,
+            )
+
+        grids_per_sample = self._normalize_batched_image_grids(image_grid_hw, batch_size)
+        expected_tokens = [sum(frame * height * width for frame, height, width in grids) for grids in grids_per_sample]
+        actual_tokens = image_attention_mask.to(torch.int64).sum(dim=1).tolist()
```
Comment (lishunyang12) [inline vllm_omni/diffusion/models/mage_flow/mage_flow_transformer.py]: Validate image token counts before denoising instead of calling .tolist() every step.

File: tests/diffusion/models/mage_flow/test_mage_flow_transformer.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        initialize_model_parallel,
+    )
+
+    os.environ.setdefault("MASTER_ADDR", "localhost")
+    os.environ.setdefault("MASTER_PORT", "29531")
+    init_distributed_environment(
+        world_size=1,
+        rank=0,
+        local_rank=0,
+        distributed_init_method="env://",
```
Comment (lishunyang12) [inline tests/diffusion/models/mage_flow/test_mage_flow_transformer.py:31]: Use Gloo and a CPU attention backend for these CPU tests.

File: vllm_omni/diffusion/models/mage_flow/pipeline_mage_flow.py (hunk 1 above)
Comment (BruceLoveDecimal) [reply vllm_omni/diffusion/models/mage_flow/pipeline_mage_flow.py]: Fixed in d3611cc21

File: vllm_omni/diffusion/models/mage_flow/mage_flow_transformer.py (hunk 2 above)
Comment (BruceLoveDecimal) [reply vllm_omni/diffusion/models/mage_flow/mage_flow_transformer.py]: Fixed in 31a4e9a96

File: tests/diffusion/models/mage_flow/test_mage_flow_transformer.py (hunk 3 above)
Comment (BruceLoveDecimal) [reply tests/diffusion/models/mage_flow/test_mage_flow_transformer.py:31]: fixed in d3611cc21

**Comment by BruceLoveDecimal** [comment]
@lishunyang12 Thanks for the review! I've addressed all the comments and re-ran the tests to guard against regressions. Could you take another look?
